man/al.1
.TH "al" 1
.SH NAME
al, al2 \- Mono Assembly Linker
.SH SYNOPSIS
.B al [option] [source-files]
.SH DESCRIPTION
AL is the Mono assembly linker. 
.PP
This linker is used to put together assemblies from a collection of
modules (.netmodule files), assembly manifest files and resources.
.PP
Do not confuse this with the 
.B monolinker,
which is a tool to reduce the size of assemblies based on the code
used. 
.PP
Use al for processing 1.0 assemblies, use al2 to process 2.0 assemblies.
.SH OPTIONS
.TP
.B @<filename>
Read response file for more options.
.TP
.B -algid:<id>
Algorithm used to hash files. The <id> must be specified in hexadecimal.
.TP
.B -base:<addr> | -baseaddress:<addr>
Base address for the library. THIS FEATURE ISN'T IMPLEMENTED.
.TP
.B -bugreport:<filename>
Create a 'Bug Report' file. THIS FEATURE ISN'T IMPLEMENTED.
.TP
.B -comp:<text> | -company:<text>
This inserts the company name into the assembly metadata.
This is equivalent to adding the [AssemblyCompany ("<text>")] attribute
into C# source code.
.TP
.B -config:<text> | -configuration:<text>
This inserts the configuration string into the assembly metadata.
This is equivalent to adding the [AssemblyConfiguration ("<text>")] 
attribute into C# source code.
.TP
.B -copy:<text> | -copyright:<text>
This inserts the copyright message into the assembly metadata.
This is equivalent to adding the [AssemblyCopyright ("<text>")] 
attribute into C# source code.
.TP
.B -c:<text> | -culture:<text>
This inserts the supported culture into the assembly metadata.
This is equivalent to adding the [AssemblyCulture ("<text>")] 
attribute into C# source code.
.TP
.B -delay | -delay+ | -delaysign | -delaysign+
The generated assembly will be delay signed. 
This is equivalent to adding the [AssemblyDelaySignAttribute (true)] 
attribute into C# source code.
.TP
.B -delay- | -delaysign-
The generated assembly will be fully signed (i.e. not delay signed).
This is the default option when signing (-keyfile or -keyname).
.TP
.B -descr:<text> | -description:<text>
This inserts a description of the assembly into the assembly metadata.
This is equivalent to adding the [AssemblyDescription ("<text>")] 
attribute into C# source code.
.TP
.B -e:<filename> | -evidence:<filename>
This embed into the assembly the specified file as assembly security
evidences.
.TP
.B -fileversion:<version>
Optional Win32 version. This overrides the normal assembly version.
.TP
.B -flags:<flags>
Assembly flags. The <flags> must be specified in hexadecimal.
.TP
.B -fullpaths
Display files using fully-qualified filenames.
.TP
.B -keyf:<filename> | -keyfile:<filename>
Strongname (sign) the output assembly using the key pair present in 
the specified strong name key file (snk). A full key pair is required
unless the delay signing option is also specified (-delay+).
This is equivalent to adding the [AssemblyKeyFile ("<text>")] 
attribute into C# source code.
.TP
.B -keyn:<text> | -keyname:<text>
Strongname (sign) the output assembly using the key pair present in 
the specified container. Delay signing isn't supported when using key
containers.
This is equivalent to adding the [AssemblyKeyName ("<text>")] 
attribute into C# source code.
.TP
.B -main:<method>
Specifies the method name of the assembly entry point.
.TP
.B -nologo
Suppress the startup banner and copyright message.
.TP
.B -out:<filename>
Output file name for the assembly manifest.
.TP
.B -prod:<text> | product:<text>
This inserts the product name into the assembly metadata.
This is equivalent to adding the [AssemblyProduct ("<text>")] 
attribute into C# source code.
.TP
.B -productv[ersion]:<text>
This inserts the product version into the assembly metadata.
This is equivalent to adding the [AssemblyInformationalVersion ("<text>")] 
attribute into C# source code.
.TP
.B -t[arget]:<target> | -target=<target>
Valid <target>s are:
.B lib | library
to create a library (.dll),
.B exe
to create a console executable (.exe), and
.B win | winexe
to create a Windows executable (.exe).
.TP
.B -template:<filename>
Specifies an assembly to get default options from.
.TP
.B -title:<text>
This inserts an assembly title into the assembly metadata.
This is equivalent to adding the [AssemblyTitle ("<text>")] 
attribute into C# source code.
.TP
.B -trade:<text> | -trademark:<text>
This inserts a trademark message into the assembly metadata.
This is equivalent to adding the [AssemblyTrademark ("<text>")] 
attribute into C# source code.
.TP
.B -v:<version> | -version:<version>
This inserts the assembly version into the assembly metadata. You can
use * to auto-generate remaining numbers.
This is equivalent to adding the [AssemblyVersion ("<text>")] 
attribute into C# source code.
.TP
.B -win32icon:<filename>
Use this icon for the output.
.TP
.B -win32res:<filename>
Specifies the Win32 resource file.
.TP
.B -? | -help
Display information about AL
.PP
.SH SEE ALSO
sn(1), monolinker(1)
.SH COPYRIGHT
Copyright (C) 2005 Novell, Inc (http://www.novell.com)
.SH MAILING LISTS
Visit http://lists.ximian.com/mailman/listinfo/mono-devel-list for details.
.SH WEB SITE
Visit: http://www.mono-project.com for details


NEWS
This is a useless commit

Release notes for Mono are hosted on the web, please see:

	http://www.mono-project.com/docs/about-mono/releases/


mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/AL.cs
//
// AL.cs: Task for assembly linker
//
// Author:
//   Marek Sieradzki (marek.sieradzki@gmail.com)
//
// (C) 2005 Marek Sieradzki
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


using System;
using System.Diagnostics;
using System.IO;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;
using Mono.XBuild.Utilities;

namespace Microsoft.Build.Tasks {
	public class AL : ToolTaskExtension {
	
		public AL ()
		{
		}
		
		[MonoTODO]
		protected internal override void AddResponseFileCommands (
						 CommandLineBuilderExtension commandLine)
		{
			commandLine.AppendSwitchIfNotNull ("/algid:", AlgorithmId);
			commandLine.AppendSwitchIfNotNull ("/baseaddress:", BaseAddress);
			commandLine.AppendSwitchIfNotNull ("/company:", CompanyName);
			commandLine.AppendSwitchIfNotNull ("/configuration:", Configuration);
			commandLine.AppendSwitchIfNotNull ("/culture:", Culture);
			commandLine.AppendSwitchIfNotNull ("/copyright:", Copyright);
			if (Bag ["DelaySign"] != null)
				if (DelaySign)
					commandLine.AppendSwitch ("/delaysign+");
				else
					commandLine.AppendSwitch ("/delaysign-");
			commandLine.AppendSwitchIfNotNull ("/description:", Description);
			if (EmbedResources != null) {
				foreach (ITaskItem item in EmbedResources) {
					string logical_name = item.GetMetadata ("LogicalName");
					if (!string.IsNullOrEmpty (logical_name))
						commandLine.AppendSwitchIfNotNull ("/embed:", string.Format ("{0},{1}", item.ItemSpec, logical_name));
					else
						commandLine.AppendSwitchIfNotNull ("/embed:", item.ItemSpec);
				}
			}
			commandLine.AppendSwitchIfNotNull ("/evidence:", EvidenceFile);
			commandLine.AppendSwitchIfNotNull ("/fileversion:", FileVersion);
			commandLine.AppendSwitchIfNotNull ("/flags:", Flags);
			if (GenerateFullPaths)
				commandLine.AppendSwitch ("/fullpaths");
			commandLine.AppendSwitchIfNotNull ("/keyname:", KeyContainer);
			commandLine.AppendSwitchIfNotNull ("/keyfile:", KeyFile);
			if (LinkResources != null)
				foreach (ITaskItem item in LinkResources)
					commandLine.AppendSwitchIfNotNull ("/link:", item.ItemSpec);
			commandLine.AppendSwitchIfNotNull ("/main:", MainEntryPoint);
			if (OutputAssembly != null)
				commandLine.AppendSwitchIfNotNull ("/out:", OutputAssembly.ItemSpec);
			//platform
			commandLine.AppendSwitchIfNotNull ("/product:", ProductName);
			commandLine.AppendSwitchIfNotNull ("/productversion:", ProductVersion);
			if (ResponseFiles != null)
				foreach (string s in ResponseFiles)
					commandLine.AppendFileNameIfNotNull (String.Format ("@{0}", s));
			if (SourceModules != null)
				foreach (ITaskItem item in SourceModules)
					commandLine.AppendFileNameIfNotNull (item.ItemSpec);
			commandLine.AppendSwitchIfNotNull ("/target:", TargetType);
			commandLine.AppendSwitchIfNotNull ("/template:", TemplateFile);
			commandLine.AppendSwitchIfNotNull ("/title:", Title);
			commandLine.AppendSwitchIfNotNull ("/trademark:", Trademark);
			commandLine.AppendSwitchIfNotNull ("/version:", Version);
			commandLine.AppendSwitchIfNotNull ("/win32icon:", Win32Icon);
			commandLine.AppendSwitchIfNotNull ("/win32res:", Win32Resource);
		}
		
		public override bool Execute ()
		{
			return base.Execute ();
		}

		protected override string GenerateFullPathToTool ()
		{
			if (!string.IsNullOrEmpty (ToolPath))
				return Path.Combine (ToolPath, ToolExe);
			return ToolLocationHelper.GetPathToDotNetFrameworkFile (ToolExe, TargetDotNetFrameworkVersion.VersionLatest);
		}

		public string AlgorithmId {
			get { return (string) Bag ["AlgorithmId"]; }
			set { Bag ["AlgorithmId"] = value; }
		}

		public string BaseAddress {
			get { return (string) Bag ["BaseAddress"]; }
			set { Bag ["BaseAddress"] = value; }
		}

		public string CompanyName {
			get { return (string) Bag ["CompanyName"]; }
			set { Bag ["CompanyName"] = value; }
		}

		public string Configuration {
			get { return (string) Bag ["Configuration"]; }
			set { Bag ["Configuration"] = value; }
		}

		public string Copyright {
			get { return (string) Bag ["Copyright"]; }
			set { Bag ["Copyright"] = value; }
		}

		public string Culture {
			get { return (string) Bag ["Culture"]; }
			set { Bag ["Culture"] = value; }
		}

		public bool DelaySign {
			get { return GetBoolParameterWithDefault ("DelaySign", false); }
			set { Bag ["DelaySign"] = value; }
		}

		public string Description {
			get { return (string) Bag ["Description"]; }
			set { Bag ["Description"] = value; }
		}

		public ITaskItem[] EmbedResources {
			get { return (ITaskItem[]) Bag ["EmbedResources"]; }
			set { Bag ["EmbedResources"] = value; }
		}

		public string EvidenceFile {
			get { return (string) Bag ["EvidenceFile"]; }
			set { Bag ["EvidenceFile"] = value; }
		}

		public string FileVersion {
			get { return (string) Bag ["FileVersion"]; }
			set { Bag ["FileVersion"] = value; }
		}

		public string Flags {
			get { return (string) Bag ["Flags"]; }
			set { Bag ["Flags"] = value; }
		}

		public bool GenerateFullPaths {
			get { return GetBoolParameterWithDefault ("GenerateFullPaths", false); }
			set { Bag ["GenerateFullPaths"] = value; }
		}

		public string KeyContainer {
			get { return (string) Bag ["KeyContainer"]; }
			set { Bag ["KeyContainer"] = value; }
		}

		public string KeyFile {
			get { return (string) Bag ["KeyFile"]; }
			set { Bag ["KeyFile"] = value; }
		}

		public ITaskItem[] LinkResources {
			get { return (ITaskItem[]) Bag ["LinkResources"]; }
			set { Bag ["LinkResources"] = value; }
		}

		public string MainEntryPoint {
			get { return (string) Bag ["MainEntryPoint"]; }
			set { Bag ["MainEntryPoint"] = value; }
		}

		[Required]
		[Output]
		public ITaskItem OutputAssembly {
			get { return (ITaskItem) Bag ["OutputAssembly"]; }
			set { Bag ["OutputAssembly"] = value; }
		}

		public string Platform {
			get { return (string) Bag ["Platform"]; }
			set { Bag ["Platform"] = value; }
		}

		public string ProductName {
			get { return (string) Bag ["ProductName"]; }
			set { Bag ["ProductName"] = value; }
		}

		public string ProductVersion {
			get { return (string) Bag ["ProductVersion"]; }
			set { Bag ["ProductVersion"] = value; }
		}

		public string[] ResponseFiles {
			get { return (string[]) Bag ["ResponseFiles"]; }
			set { Bag ["ResponseFiles"] = value; }
		}

		public ITaskItem[] SourceModules {
			get { return (ITaskItem[]) Bag ["SourceModules"]; }
			set { Bag ["SourceModules"] = value; }
		}

		public string TargetType {
			get { return (string) Bag ["TargetType"]; }
			set { Bag ["TargetType"] = value; }
		}

		public string TemplateFile {
			get { return (string) Bag ["TemplateFile"]; }
			set { Bag ["TemplateFile"] = value; }
		}

		public string Title {
			get { return (string) Bag ["Title"]; }
			set { Bag ["Title"] = value; }
		}

		protected override string ToolName {
			get { return "al.exe"; }
		}

		public string Trademark {
			get { return (string) Bag ["Trademark"]; }
			set { Bag ["Trademark"] = value; }
		}

		public string Version {
			get { return (string) Bag ["Version"]; }
			set { Bag ["Version"] = value; }
		}

		public string Win32Icon {
			get { return (string) Bag ["Win32Icon"]; }
			set { Bag ["Win32Icon"] = value; }
		}

		public string Win32Resource {
			get { return (string) Bag ["Win32Resource"]; }
			set { Bag ["Win32Resource"] = value; }
		}
	}
}

